sandbox/samples/cpp/alembicassembly/alembicassembly.cpp

```diff
@@ -0,0 +1,1525 @@
+#include "foundation/math/vector.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/autoreleaseptr.h"
+#include "foundation/utility/job/iabortswitch.h"
+#include "foundation/utility/string.h"
+
+// appleseed.main headers.
+#include "main/dllvisibility.h"
+
+// Alembic headers.
+#include <Alembic/AbcCoreFactory/All.h>
+#include <Alembic/AbcGeom/All.h>
+
+// Standard headers.
+#include <cstddef>
+#include <string>
+#include <vector>
+#include <algorithm>
+
+namespace
+{
+    // shortcuts to save space
+    namespace asf = foundation;
+    namespace asr = renderer;
+
+    namespace Abc        = Alembic::Abc;
+    namespace AbcCoreFac = Alembic::AbcCoreFactory;
+    namespace AbcGeom    = Alembic::AbcGeom;
+    namespace AbcUtil    = Alembic::Util;
+
+    typedef Alembic::Abc::index_t  AbcIndex_t;
+    typedef Alembic::Abc::chrono_t AbcChrono_t;
+
+    // roll into given matrix stack and return the flatten one
+    asf::Matrix4d flatten_xform(const std::vector<asf::Matrix4d>& mtx_stack)
+    {
+        // the matrix we will return
+        asf::Matrix4d out_mtx = asf::Matrix4d::identity();
+
+        // multiply the each matrix of the stack from bottom to top
+        for (const asf::Matrix4d& mtx : mtx_stack)
+        {
+            out_mtx = mtx * out_mtx;
+        }
+
+        return out_mtx;
+    }
+
+    // roll into given xform sequence stack and return the flatten one
+    asr::TransformSequence flatten_xform_seq(const std::vector<asr::TransformSequence>& xform_seq_stack)
+    {
+        assert(xform_seq_stack.size() > 0);
+
+        if (xform_seq_stack.size() == 1)
+        {
+            // one item, we don't need to flatten anything
+            return xform_seq_stack.back();
+        }
+        else
+        {
+            // xform sequence we will return
+            asr::TransformSequence out_xform_seq;
+
+            for (const asr::TransformSequence& xform_seq : xform_seq_stack)
+            {
+                out_xform_seq = xform_seq * out_xform_seq;
+            }
+
+            out_xform_seq.optimize();
+            out_xform_seq.prepare();
+
+            return out_xform_seq;
+        }
+    }
+
+    // return if given time samples are linearly time spaced between start_time
+    // and end_time.
+    const bool are_linearly_sampled(const AbcChrono_t start_time,
+                                    const AbcChrono_t end_time,
+                                    const std::vector<AbcChrono_t>& samples)
+    {
+        assert(samples.size() > 1);
+
+        // compute time supposed to be between each samples if samples are
+        // linearly sampled.
+        const AbcChrono_t increm = (end_time - start_time) / (samples.size()-1);
+
+        // start sample
+        AbcChrono_t t_accum = samples[0];
+
+        // get each sample time and compare with a linearly sampled time
+        for (const AbcChrono_t t : samples)
+        {
+            if (asf::feq(t, t_accum))
```

Comment:
Are you sure about the logic here? Shouldn't it be `!asf::feq()`?

Reply:
Done.
Good catch! This code is not used yet as its purpose was to prevent the user of bad sampling. I just don't know yet _when_ I will check and report.